sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js

```diff
@@ -458,8 +468,276 @@ Ext.define('SlateTasksTeacher.controller.Dashboard', {
         });
     },
 
+    onAddGoogleAttachmentClick: function() {
+        var me = this,
+            taskEditor = me.getTaskEditor(),
+            token = Ext.util.Cookies.get('googleAppsToken', '/');
+
+        if (!taskEditor.getTask().phantom) {
+            Ext.Msg.alert('Error', 'Unable to add attachments while editing&hellip; Please create a new task.');
+            return;
+        }
+        // check for google token in cookies
+        gapi.load('auth:picker:client', {
+            callback: function() {
+                // TODO: check if token is any good
+                if (token) {
+                    return me.doOpenFilePicker(token);
+                }
+
+                me.doGoogleAuthentication();
```

Comment:
@nbey doOpenFilePicker should get passed as a callback to this method rather than being hard-coded into the chain of functions that follows authentication

We're going to want to break out all these functions to a sharable singleton class in the slate-cbl package so that other UIs (like the student submission UI we'll implement by the end of this project) can use the same code. Think about how to implement a clean, resuable set of methods there that could be called from here and chained together here through callback parameters